classes/form/template.php

```diff
@@ -0,0 +1,114 @@
+ * @copyright  2020 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace quizaccess_seb\form;
+
+use CFPropertyList\CFPropertyList;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once(__DIR__ . '/../../vendor/autoload.php');
+
+
+class template extends \core\form\persistent {
+
+    /** @var string Persistent class name. */
+    protected static $persistentclass = 'quizaccess_seb\\template';
+
+    /**
+     * @inheritdoc
+     */
+    protected function definition() {
+        $mform = $this->_form;
+
+        $mform->addElement('text', 'name', get_string('name', 'quizaccess_seb'));
+        $mform->addRule('name', get_string('required'), 'required', null, 'client');
+        $mform->setType('name', PARAM_TEXT);
+
+        $mform->addElement('textarea', 'description', get_string('description', 'quizaccess_seb'));
+        $mform->setType('description', PARAM_TEXT);
+
+        if ($this->get_persistent()->get('id')) {
+            $mform->addElement('textarea', 'content', get_string('content', 'quizaccess_seb'), ['rows' => 20, 'cols' => 60]);
+            $mform->addRule('content', get_string('uploadnofilefound'), 'required');
+        } else {
+            $mform->addElement('filepicker', 'content', get_string('content', 'quizaccess_seb'));
+            $mform->addRule('content', get_string('uploadnofilefound'), 'required');
+        }
+
+        $mform->addElement('selectyesno', 'enabled', get_string('enabled', 'quizaccess_seb'));
+        $mform->setType('enabled', PARAM_INT);
+
+        $this->add_action_buttons();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    protected function filter_data_for_persistent($data) {
+        // Uploading a new template file.
+        if (empty($this->get_persistent()->get('id'))) {
+            $files = $this->get_draft_files('content');
+            if ($files) {
+                $file = reset($files);
+                $data->content = $file->get_content();
+            } else {
+                // No file found. Remove content data and let persistent to return an error.
+                unset($data->content);
+            }
+        }
+
+        return parent::filter_data_for_persistent($data);
+    }
+
+    /**
+     * Extra validation.
+     *
+     * @param  \stdClass $data Data to validate.
+     * @param  array $files Array of files.
+     * @param  array $errors Currently reported errors.
+     * @return array of additional errors, or overridden errors.
+     */
+    protected function extra_validation($data, $files, array &$errors) {
+        $newerrors = [];
+
+        // Check name.
+        if (empty($data->name)) {
+            $newerrors['name'] = get_string('namerequired', 'quizaccess_seb');
+        }
+
+        // Check submitted template as a text.
+        if (!empty($this->get_persistent()->get('id'))) {
+            $plist = new CFPropertyList();
+            try {
+                $plist->parse($data->content);
```

Comment:
With full Moodle debugging enabled, 

CFProperyList Parse throws a warning if $data->content is empty
`Warning: DOMDocument::loadXML(): Empty string supplied as input`

If the field is not valid XML it also warns
`Warning: DOMDocument::loadXML(): Start tag expected, '<' not found in Entity`

We have some tests which over this regarding the config key generation, test_config_key_not_generated_with_bad_xml

Reply:
I have added custom error handler.